### ISSUE TITLE
Feature Requests: 2020 - IB, RDMA, Dedup, Compression, Caching

### ISSUE THREAD

**Issue by jwest75674** [lizardfs/lizardfs#868]
Hello, Hope all is well!

After reading through #278, I figured I would bump the issue as well as request a few other features I would love to eventually see.

I have been working around the lack of direct IB support from lizardfs by using Infiniband + ISER via ISCSI exporting each block, and running both (multiple chunkserver processes, one per disk) chunk servers and master from a single machine.
Basically, no lizardfs IB support has been worked around by using IB to first make all devices appear local to the server, then setting up lizardfs, but it makes for one hell of a confusing setup process trying to manage it all. Then you layer Multipath, VDO, and caching, and good lord is it a nightmare to keep all the WWNs and UUIDs in order.

Since I'm voicing feature requests:
VDO is the block level deduplication solutions I've settled on. For my use case it not only recovers my EC{} lost storage, but deduplication actually saves me an additional 15%. (My use case being 100% 1GB .gz files from the common crawl corpus. 
Again, this is a layer of complexity in management during setup which would be awesome to see asn an integration // native option. Transparent disk deduplication and compression.
--> Anyone following in these steps down the road, consider exporting the VDO active block storage via ISCSI, instead of running VDO on the master/chunkserver machine, to ensure each node in the cluster handles the processing and ram requirements of deduplication and compression, instead of centralizing this computing on the master server.

Lastly, Tiering or Caching. This has a popular issue or two, #524 , #710 both pop to mind. 

I have three m.2 NVME drives, currently sitting around, as for the life of my I have not been able to find a method to get the above working with any form of caching.
As Lizardfs doesn't support storage teiring, and my preference is EC{}, I turned to Caching.

The caches I've tried (and failed for one reason or another) are dm-writeboost, EnhanceIO (the still maintained fork), bcache. Each wants to work on blocks, meaning trying to repeat a task nearly 30 times across each block. Would love to see a method to leverage 1.5TB of screaming fast NVME to speed up lizardfs. (Open to ideas if anyone has anything..?)

I even considered IOnotify, to simply write to the SSD's, and automatically move the files to lizardfs, but this has it's own issues (directory recursion, as an example). 

Ideal setup would be to "assign" a device, directory, etc. as an SSD write cache. and otherwise not have to worry about it.

Deduplication, compression, caching, Infiniband. <3

**Comment by szycha76** [comment]
Yeah, even file-level deduplication could be a game-changer in certain scenarios.

Some stupid advice crossing my mind related to caching, tiering, compression & deduplication would be: put your chunkservers on ZFS with compression & deduplication turned on. Put your  M.2 drives as ZIL & cache devices in your zpools - and you *could* get some of your issues resolved. Keep in mind, however, you'll need tons of RAM for ZFS deduplication, though.

**Comment by jwest75674** [comment]
Having learned more about how MFS works, would chunk level deduplication be within grasp?

We're then talking about stripping the chunk header, and generating a hash for each chunk, looking for identical hashes. Then use of a link seems like it would not be far from practical.

**Comment by szycha76** [comment]
Should be, since `lizardfs makesnapshot` creates distinct files that share chunks. How much do you expect to gain with chunk size at default level of 64 MiB, though?

Yet chunk-level deduplication seems a low-hanging fruit and could be a great *selling point* over your direct competition, right @aNeutrino?
